plugins/analytics/public/virustotal_api.py

```diff
@@ -0,0 +1,105 @@
+        try:
+            response = None
+            if isinstance(observable, Hostname):
+                params = BASE_URL_PARAMS
+                params['resource'] = observable.value
+                params['apikey'] = api_key
+                response = urllib.urlopen('%s?%s' % (BASE_URL_URL, urllib.urlencode(params)))
+            elif isinstance(observable, Ip):
+                params = BASE_IP_PARAMS
+                params['ip'] = observable.value
+                params['apikey'] = api_key
+                response = urllib.urlopen('%s?%s' % (BASE_IP_URL, urllib.urlencode(params)))
+
+            if response.code == 200:
+                # self.last_query['time'] = datetime.now()
+                # self.last_query['count'] += 1
+                return response.read()
+            else:
+                return None
+        except Exception as e:
+            print 'Exception while getting ip report %s' % e.message
+            return None
+
+
+class VirusTotalQuery(OneShotAnalytics, VirustotalApi):
+    default_values = {
+        'name': 'Virustotal',
+        'description': 'Perform a Virustotal query.',
+    }
+
+    ACTS_ON = ['Ip', 'Hostname']
+
+    @staticmethod
+    def analyze(observable, results):
+        links = set()
+        response = VirustotalApi.fetch(observable, results.settings['virutotal_api_key'])
+        json_result = json.loads(response)
+        results.update(raw=json.dumps(json_result, sort_keys=True, indent=4, separators=(',', ': ')))
+        result = dict()
+        result['raw'] = json.dumps(json_result, sort_keys=True, indent=4, separators=(',', ': '))
+
+        if isinstance(observable, Ip):
+            # Parse results for ip
+            if 'as_owner' in json_result and json_result['as_owner'] is not None:
+                result['Owner'] = json_result['as_owner']
+                o_isp = Company.get_or_create(name=json_result['as_owner'])
+                links.update(observable.active_link_to(o_isp, 'hosting', VT_QUERY))
+
+            if 'detected_urls' in json_result and json_result['detected_urls'] is not None:
+                result['detected_urls'] = json_result['detected_urls']
+                for detected_url in json_result['detected_urls']:
+                    o_url = Url.get_or_create(value=detected_url['url'])
+                    links.update(o_url.active_link_to(o_url, 'hostname', VT_QUERY))
+
+        elif isinstance(observable, Hostname):
+            if 'permalink' in json_result and json_result['permalink'] is not None:
+                result['permalink'] = json_result['permalink']
+
+            if 'positives' in json_result and json_result['positives'] is not None:
+                result['positives'] = json_result['positives']
+            else:
+                result['positives'] = 0
+
+            if 'total' in json_result and json_result['total'] is not None:
+                result['total'] = json_result['total']
+
+        for context in observable.context:

```

Comment:
Can the context change over time? If the answer is yes, then it might be worth keeping historical records (also since this is a OneShot analytics I don't think it will generate too much noise). Take a look at #119 (which I will merge in a few hours) which changes (improves!) the logic of how context deduplication works.